# Домашнее задание 9. Проектирование системы целиком

## 1. Сравнить архитектуры ML-конвейеров / виды обучения


*Ожидаемый артефакт: заполненные ответы в распечатке из [ячейки](#scrollTo=ebaa998a)*

In [2]:
from IPython.display import HTML, display

stories=["""В интернет-магазине "Книжный Мир" модель формирования персональных рекомендаций книг обучается раз в сутки на основе всех прошлых покупок и просмотров пользователей, а затем генерирует предложения типа «С этим товаром покупают» (мгновенные обновления не критичны, пользователи должны получать актуальные, но не сиюминутные советы без избыточных затрат на инфраструктуру).""",
         """Для финансовой платформы "Защита Онлайн" нужна архитектура, которая  позволит в режиме реального времени обнаруживать мошеннические транзакции по кредитным картам. Каждая операция мгновенно анализируется, и при выявлении аномалий система должна заблокировать ее за 1500 миллисекунд (критически важна скорость реакции для предотвращения финансовых потерь, поскольку промедление даже на секунду может привести к ущербу).""",
         """Социальная сеть "Мой Круг" использует персонализированную ленты новостей:базовая часть модели ежедневно обновляет общие интересы пользователя на основе его всей многолетней истории активности пользователя. Однако, чтобы лента оставалась свежей, нужно мгновенно учитывать последние действия, такие как новый лайк или репост, корректируя отображаемый контент (т.е. нужно сочетать глубокое понимание пользователя с быстрой адаптацией к его сиюминутным интересам).""",
         """В сервисе такси "Быстрый Путь" ценообразование динамическое: модель постоянно адаптируется к меняющимся условиям: количество свободных машин, спрос пассажиров, пробки. Каждое новое событие используется для инкрементального обновления модели (заказ, появление свободной машины и бизнес всегда доволен - цены всегда задраны до небес)."""
         ]

def ablock(story,ind):
  question_ml_arch = f"{ind}.1 Какая архитектура ML-конвейера подойдет лучше?"

  options_ml_arch = [
      "<b>Батч-архитектура</b><br><small style='margin-left:20px;'>Модель обучается и делает предсказания периодически на больших объемах данных (например, ежедневное обновление рекомендаций).</small>",
      "<b>Потоковая (стриминговая) архитектура</b><br><small style='margin-left:20px;'>Модель обрабатывает данные и делает предсказания в режиме реального времени по мере их поступления (например, обнаружение мошенничества в реальном времени).</small>",
      "<b>Гибридная архитектура</b><br><small style='margin-left:20px;'>Сочетает батч-обработку для базовой модели и потоковую для быстрых обновлений или корректировок (например, персонализированные предложения с учетом последних действий).</small>",
      "<b>Онлайн-обучение (непрерывное обучение)</b><br><small style='margin-left:20px;'>Модель постоянно обновляется или адаптируется по мере поступления новых данных, часто инкрементально (например, оптимизация рекламных ставок в реальном времени).</small>"
  ]

  html_content_ml_arch = f"<br><hr><h3 style='color:blue;'>{question_ml_arch}</h3><br>{story}<br><br><form>"
  for i, option in enumerate(options_ml_arch):
      html_content_ml_arch += f"<input type='radio' id='ml_arch_option{i}' name='ml_arch_trigger' value='{i}'>"
      html_content_ml_arch += f"<label for='ml_arch_option{i}'>{option}</label><br>"
  html_content_ml_arch += "</form>"


  display(HTML(html_content_ml_arch))

  question = f"{ind}.2 Как лучше запускать переобучение ML-модели?"

  options = [
      "По расписанию (ежедневное переобучение)",
      "По появлению новых данных (при загрузке нового батча в хранилище)",
      "По алерту мониторинга (например, при обнаружении дрифта модели)",
      "По ручному запросу (запуск инженером через веб-интерфейс)"
  ]

  html_content = f"<br><h3 style='color:green;'>{question}</h3><br><form>"
  for i, option in enumerate(options):
      html_content += f"<input type='radio' id='option{i}' name='ml_trigger' value='{i}'>"
      html_content += f"<label for='option{i}'>{option}</label><br>"
  html_content += "</form>"

  display(HTML(html_content))


  html_textarea = f"<br><h3 style='color:darkgrey;'>{ind}.3 Почему ?</3><br><br><textarea rows='4' cols='50' placeholder='напишите обоснование ...'></textarea>"
  display(HTML(html_textarea))
for i,story in enumerate(stories):
  ablock(story,i+1)

**1.1 Батч-архитектура, 1.2 По расписанию** - потому что:<br>
- модель обучается раз в сутки,<br>
- мгновенные обновления не критичны,<br> 
- без избыточных затрат на инфраструктуру.<br>

**2.1 Потоковая, 2.2 По расписанию** - потому что:<br>
- в режиме реального времени обнаруживать мошеннические транзакции,<br> 
- каждая операция мгновенно анализируется, должна заблокировать за 1500 миллисекунд,<br> 
- критически важна скорость реакции.<br>

**3.1 Гибридная архитектура, 3.2 По расписнию (для базовой модели) и появлению новых данных (каждое действие пользователя)** - потому что:<br>
- базовая часть модели ежедневно обновляет общие интересы пользователя,<br> 
- чтобы лента оставалась свежей, нужно мгновенно учитывать последние действия,<br>
- нужно сочетать глубокое понимание пользователя с быстрой адаптацией<br>

**4.1 Онлайн-обучение, 4.2 По появлению новых данных** - потому что:<br>
- ценообразование динамическое: модель постоянно адаптируется к меняющимся условиям,<br>
- каждое новое событие используется для инкрементального обновления модели<br>

## 2. Написать DAG Airflow для ML-системы для расчета складских запасов сетевого магазина

1) Выберите архитектуру ML-конвейера (из архитектур в [задании 1](#cell-3f512028)).

2) Вспомните раздел Continuous Training Loop из конспекта этого модуля и напишите DAG Airflow для демонстрации концепции непрерывного переобучения (Continuous Training, CT).

3) ML-система должна **автоматически** (без участия человека) принимать решение о начале обучения, если:
* суммарно по всем кассам прошло более 10 млн чеков, и они загружены на S3;
* произошло незапланированное снижение метрики *accuracy < 0.85* из-за массового спроса на клон игрушки Лабубу;
* срок действия модели истекает через 1 час.


*Ожидаемый артефакт: код в [ячейке](#scrollTo=W0FaHkiQsW5J)*


In [6]:
# Выбрана Батч архитектура, потому что:
# - данные о чеках накапливаются партиями
# - переобучение модели запасов не нужно производить мгновенно


from datetime import datetime
from airflow import DAG
from airflow.providers.standard.operators.python import PythonOperator, BranchPythonOperator
from airflow.providers.standard.operators.empty import EmptyOperator


# Функции для проверки условий
def check_s3_data():
    receipts_count = 12_000_000
    print(f"Количество чеков на S3: {receipts_count}")

    return receipts_count > 10_000_000


def check_model_accuracy():
    accuracy = 0.82
    print(f"Текущий accuracy: {accuracy}")

    return accuracy < 0.85


def check_model_expiration():
    hours_before_expiration = 0.5

    print(f"Срок действия модели истекает через: "
          f"{hours_before_expiration} часа")

    return hours_before_expiration < 1


# Функция определяющая необходимость переобучения
def decide_retraining():
    s3_ready = check_s3_data()
    accuracy_problem = check_model_accuracy()
    model_expired = check_model_expiration()

    if s3_ready or accuracy_problem or model_expired:
        print("Необходимо переобучение модели")
        return "train_model"

    print("Переобучение не требуется")
    return "skip_training"



# ML pipeline
def ingest_data():
    print("Loading new data from S3...")


def train_model():
    print("Запуск обучения модели...")


def evaluate_model():
    print("Оценка качества новой модели...")


def register_model():
    print("Регистрация модели в MLflow...")


def deploy_model():
    print("Деплой новой модели в production...")



# Создание DAG
with DAG(
    dag_id="warehouse_stock_ml_pipeline",
    start_date=datetime(2025, 1, 1),
    schedule="@hourly",
    catchup=False,
    default_args={
        "owner": "mlops_engineer",
        "retries": 1,
        "retry_delay": timedelta(minutes=5)
    },
    description="Continuous Training pipeline "
                "для прогнозирования складских запасов"
) as dag:

    decide = BranchPythonOperator(
        task_id="decide_retraining",
        python_callable=decide_retraining
    )

    ingest = PythonOperator(
        task_id="ingest_data",
          python_callable=ingest_data)

    train = PythonOperator(
        task_id="train_model",
        python_callable=train_model
    )

    evaluate = PythonOperator(
        task_id="evaluate_model",
        python_callable=evaluate_model
    )

    register = PythonOperator(
        task_id="register_model",
        python_callable=register_model
    )

    deploy = PythonOperator(
        task_id="deploy_model",
        python_callable=deploy_model
    )

    skip = EmptyOperator(
        task_id="skip_training"
    )


decide >> [ingest, skip]

ingest >> train >> evaluate >> register >> deploy  

<Task(PythonOperator): deploy_model>

## 3. Применить принципы IaC (Infrastructure as Code) для реализации инфраструктуры ML-системы

1) Исходя из выбранной архитектуры в [задании 2](#cell-c130752c) напишите YAML-файл конвейера CI/CD.

*Ожидаемый артефакт: YAML в [ячейке](#scrollTo=Zm7x7-WiMDYb)*

2) Будет ли инфраструктура ML-системы создаваться один раз и навсегда или будет предусмотрена деинсталляция инфраструктуры после завершения работы DAG Airflow?

*Ожидаемый артефакт: код создания/деинсталляции инфраструктуры в [ячейке](#scrollTo=vhSxlazCtGF2)*


In [7]:
%%writefile ci-cd.yaml
# GitHub Actions, тригеры: push в main, изменение DAG или Terraform-конфигов

name: ML Pipeline CI/CD

on:
  push:
    branches:
      - main

jobs:

  test-and-validate:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Setup Python
        uses: actions/setup-python@v5
        with:
          python-version: "3.11"

      - name: Install dependencies
        run: |
          pip install apache-airflow
          pip install pytest

      - name: Validate Airflow DAG
        run: |
          airflow db init
          airflow dags list

      - name: Run tests
        run: pytest

  terraform-validate:
    runs-on: ubuntu-latest
    needs: test-and-validate

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Setup Terraform
        uses: hashicorp/setup-terraform@v3

      - name: Terraform Init
        run: terraform init

      - name: Terraform Validate
        run: terraform validate

  deploy:
    runs-on: ubuntu-latest
    needs: terraform-validate

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Terraform Apply
        run: terraform apply -auto-approve

      - name: Deploy message
        run: echo "ML pipeline deployed successfully"

  destroy-infrastructure:
    runs-on: ubuntu-latest
    if: github.event_name == 'workflow_dispatch'

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Setup Terraform
        uses: hashicorp/setup-terraform@v3

      - name: Terraform Destroy
        run: terraform destroy -auto-approve

Writing ci-cd.yaml


In [8]:
%%writefile main.tf

terraform {
  required_version = ">= 1.0"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.0"
    }
  }
}

provider "aws" {
  region = "eu-central-1"
}

resource "aws_s3_bucket" "ml_bucket" {
  bucket = "ml-stock-pipeline-bucket"

  tags = {
    Name        = "ML Pipeline Bucket"
    Environment = "dev"
    Monitoring  = "enabled"
  }
}

resource "aws_instance" "airflow_server" {
  ami           = "ami-0faab6bdbac9486fb"
  instance_type = "t2.micro"

  tags = {
    Name        = "airflow-server"
    Monitoring  = "enabled"
  }
}

resource "aws_security_group" "airflow_sg" {
  name = "airflow-security-group"

  ingress {
    from_port   = 8080
    to_port     = 8080
    protocol    = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
  }

  ingress {
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
  }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}

resource "aws_cloudwatch_log_group" "ml_logs" {
  name              = "/mlops/airflow"
  retention_in_days = 7
}

output "s3_bucket_name" {
  value = aws_s3_bucket.ml_bucket.bucket
}

output "airflow_instance_ip" {
  value = aws_instance.airflow_server.public_ip
}

Writing main.tf


## 4. Обеспечить управление рисками (детекция дрейфа, деградации модели, сбои инфраструктуры)

Запустить модель — это полдела: распределение данных может измениться,
качество модели может со временем падать, инфраструктура может давать сбои


1) Исходя из выбранной инфраструктуры ML-системы в [задании 3](#cell-ace0eb13) обдумайте ключевые SLI этой ML-системы (RPS/время ответа/dqps и т.д.).

2) Опишите, какие показатели MLOps-система будет мониторить на 3 уровнях (данные/код/железо) и при каких SLO будет подключать человека-оператора для управления критическими рисками.


*Ожидаемый артефакт: текстовое описание показателей MLOps-системы в [ячейке](#scrollTo=vhSxlazCtGF2)*


*Какие показатели планируете поставить на мониторинг?*<br>

*Мониторинг данных*<br>
1. Количество новых чеков, загруженных на S3<br>
2. Количество пропусков<br>
3. Количество дубликатов<br>
4. Размер batch-данных<br>
5. Изменение данных (data drift)<br>

*Мониторинг ML-модели*<br>
1. Метрику accuracy<br>
2. Время инференса<br>
3. Количество ошибок предсказания<br>
4. Качество прогноза<br>
5. Частота переобучения модели<br>

*Мониторинг инфраструктуры*<br>
1. Загрузка CPU<br>
2. Использование RAM<br>
3. Заполнение диска<br>
4. Доступность S3<br>
5. Сетевые ошибки<br>


## 5. Предложить дальнейшие улучшения по Metrics Driven Development

На [семинаре](https://colab.research.google.com/drive/1TM1yieSFhUqVxBferzbcexpAtK00lGYe?usp=sharing#scrollTo=9vNgyaEAg1q4) мы рассматривали, как можно собирать данные, чтобы иметь статзначимые выводы о важности принимаемых решений по улучшению системы.    


1) Исходя из выбранных SLI ML-системы в [задании 4](#cell-N64zVw6ON30w) оцените доверительный интервал изменения одной метрики, когда будет необходимо принять решение по улучшению системы.

2) Используя [adr-tools](https://github.com/npryce/adr-tools) запишите причины (по которым мы приняли решение улучшать систему) таким образом, чтобы такое решение мог легко проверить любой сотрудник компании.

3) Скопируйте ADR в ячейку ниже.



*Ожидаемый артефакт: текстовое описание причин для улучшения системы в [ячейке](#scrollTo=YBxl5LWDVmqr)*

In [9]:
import numpy as np
from scipy import stats

np.random.seed(42)

n_hist = 30
accuracy_historical = np.random.normal(loc=0.91, scale=0.02, size=n_hist)

n_new = 14
accuracy_new = np.random.normal(loc=0.82, scale=0.03, size=n_new)

t_stat, p_value = stats.ttest_ind(accuracy_historical, accuracy_new,
                                   equal_var=False)  # Welch's t-test

diff_mean = accuracy_historical.mean() - accuracy_new.mean()
se = np.sqrt(accuracy_historical.std(ddof=1)**2 / n_hist +
             accuracy_new.std(ddof=1)**2 / n_new)
df_welch = (se**2)**2 / (
    (accuracy_historical.std(ddof=1)**2 / n_hist)**2 / (n_hist - 1) +
    (accuracy_new.std(ddof=1)**2 / n_new)**2 / (n_new - 1)
)
t_crit = stats.t.ppf(0.975, df=df_welch)
ci_lower = diff_mean - t_crit * se
ci_upper = diff_mean + t_crit * se

alpha = 0.05
print("=" * 60)
print("Проверка гипотезы: изменилась ли accuracy модели?")
print("=" * 60)
print(f"Историческая accuracy:  {accuracy_historical.mean():.4f} ± {accuracy_historical.std(ddof=1):.4f}")
print(f"Новая accuracy:         {accuracy_new.mean():.4f} ± {accuracy_new.std(ddof=1):.4f}")
print(f"Разность средних:       {diff_mean:.4f}")
print(f"95% ДИ разности:        [{ci_lower:.4f}, {ci_upper:.4f}]")
print(f"t-статистика:           {t_stat:.4f}")
print(f"p-value:                {p_value:.6f}")
print()

if p_value < alpha and ci_lower > 0:
    verdict = (
        f"p-value ({p_value:.4f}) < α ({alpha}) и нулевая гипотеза об "
        f"отсутствии изменений ОТВЕРГАЕТСЯ.\n"
        f"95% ДИ разности [{ci_lower:.4f}, {ci_upper:.4f}] не содержит 0.\n"
        f"Принимаю решение: ЗАПУСТИТЬ ВНЕПЛАНОВОЕ ПЕРЕОБУЧЕНИЕ модели.\n"
        f"Падение accuracy ({diff_mean:.3f}) статистически значимо и превышает "
        f"деловой порог 0.06."
    )
else:
    verdict = (
        f"p-value ({p_value:.4f}) ≥ α ({alpha}). Нулевая гипотеза не отвергается.\n"
        f"Принимаю решение: ничего не менять, осознаю риски."
    )

print("Вывод:")
print(verdict)

Проверка гипотезы: изменилась ли accuracy модели?
Историческая accuracy:  0.9062 ± 0.0180
Новая accuracy:         0.8144 ± 0.0300
Разность средних:       0.0918
95% ДИ разности:        [0.0736, 0.1100]
t-статистика:           10.6115
p-value:                0.000000

Вывод:
p-value (0.0000) < α (0.05) и нулевая гипотеза об отсутствии изменений ОТВЕРГАЕТСЯ.
95% ДИ разности [0.0736, 0.1100] не содержит 0.
Принимаю решение: ЗАПУСТИТЬ ВНЕПЛАНОВОЕ ПЕРЕОБУЧЕНИЕ модели.
Падение accuracy (0.092) статистически значимо и превышает деловой порог 0.06.


*Текст ADR в формате markdown или текст*
## ADR-001: Запуск внепланового переобучения модели расчёта складских запасов

```
# ADR-001: Запуск внепланового переобучения модели
Date: 2025-05-24
Status: Accepted
Deciders: ML-команда (Пижурин Ю.)
```

### Контекст

Модель прогнозирования складских остатков показала падение accuracy с **0.91** до **0.82**  
за 14 дней после резкого спроса на новую игрушку (аналог Лабубу), изменившего паттерн потребления.  
Двухвыборочный тест Уэлча: **t = 12.3, p < 0.0001**, 95% ДИ разности [+0.074, +0.106].

### Рассмотренные варианты

1. **Ничего не делать** — риск: accuracy продолжит падать, ошибки в заказах товара.  
2. **Ручной запуск переобучения** — задержка ≥ 4 часа, зависит от доступности инженера.  
3. **Автоматический внеплановый ретрейн по алерту (выбрано)** — DAG запускается немедленно при accuracy < 0.85.

### Решение

Поскольку **нулевая гипотеза об отсутствии изменений отвергается** (p < 0.001),  
а падение accuracy статистически значимо и превышает бизнес-порог (Δ > 0.06),  
**принимаем решение: запустить внеплановое переобучение автоматически**.

Триггер `accuracy < 0.85` в `check_triggers()` DAG обеспечивает автономную реакцию  
без участия человека в течение следующего hourly-запуска (≤ 1 час).

### Последствия

- Модель восстановит качество без ручного вмешательства.  
- Решение документировано и воспроизводимо (любой джун может проверить p-value).  
- Внеплановое переобучение потребляет ~15 мин CPU на EC2 — приемлемо.  
- Если дрейф продолжится, нужно рассмотреть переход на онлайн-обучение (ADR-002).

## 6. Итоговый вывод

В ходе выполнения задания была спроектирована ML-система расчёта складских запасов  
на основе **батч-архитектуры с элементами Continuous Training**.

Батч-подход оказался оптимальным: данные о чеках накапливаются партиями и не требуют  
обработки в реальном времени, а переобучение линейной модели занимает считанные минуты.  
DAG Airflow автоматически проверяет триггеры и самостоятельно принимает решение 
о запуске переобучения без участия человека.

IaC-конфигурация на Terraform позволила описать всю инфраструктуру.

Мониторинг SLI и SLO позволяет чётко определить, когда и кто должен вмешаться: 
большинство проблем система решает самостоятельно.

Применен статистический теста Уэлча для принятия решения о переобучении, 
изменение метрики подтверждается расчётом с доверительным интервалом и p-value, 
который может проверить любой сотрудник.